[Frontend] Respect Chat Completion parallel_tool_calls param

bbrowning opened the pull request:
## Purpose

Our Chat Completions API exposed a `parallel_tool_calls` parameter but didn't actually use that to control the number of tool calls returned. This fixes that, so that if a user sets `parallel_tool_calls` to `False` then we'll only return a single tool call for streaming or non-streaming responses instead of allowing more than one.

The default for this parameter in the Chat Completions API spec is `True`, and the previous behavior of vLLM acted like this was set to `True`. So, even though in this change the default in the code changes from `False` to `True`, from a user's point-of-view and backwards-compatibility perspective the default has not changed. We were ignoring the value entirely before and acting like it was always set to `True`.

This does not attempt to stop generation once the first tool call is generated. Instead, it simply drops any tool calls after the first when sending the response back to the client if `parallel_tool_calls` is `False`. Figuring out how or if we could stop the generation early to avoid wasting GPU cycles is above my current knowledge of the internals here.

I adjusted the code comments and documentation to remove references to `parallel_tool_calls` not being implemented, since it now is.

Fixes #9451 

## Test Plan

I added an addition test in the existing `tests/tool_use/test_parallel_tool_calls.py` that ensures when a client sends `parallel_tool_calls=False` that it only receives a single tool call in the response. The test checks this for both streaming and non-streaming clients.

I ran this locally with a couple of models to ensure things pass. The entire test suite uses more models, including ones I cannot easily run locally like llama4, but I believe the subset I ran this with is enough to verify the behavior here since the change is not model-specific.

```
pytest tests/tool_use/test_parallel_tool_calls.py --models mistral llama3.2
```

## Test Result

The new test as well as existing parallel tool call tests pass with the mistral and llama3.2 model configurations I tested.

```
tests/tool_use/test_parallel_tool_calls.py ssssssss.ssssss...ssssssssssss                [100%]
```

bbrowning committed:
[Frontend] Respect Chat Completion parallel_tool_calls param

Our Chat Completions API exposed a `parallel_tool_calls` parameter but
didn't actually use that to control the number of tool calls returned.
This fixes that, so that if a user sets `parallel_tool_calls` to `False`
then we'll only return a single tool call for streaming or non-streaming
responses instead of allowing more than one.

The default for this parameter in the Chat Completions API spec is
`True`, and the previous behavior of vLLM acted like this was set to
`True`. So, even though in this change the default in the code changes
from `False` to `True`, from a user's point-of-view and
backwards-compatibility perspective the default has not changed. We were
ignoring the value entirely before and acting like it was always set to
`True`.

This does not attempt to stop generation once the first tool call is
generated. Instead, it simply drops any tool calls after the first when
sending the response back to the client if `parallel_tool_calls` is
`False`. Figuring out how or if we could stop the generation early to
avoid wasting GPU cycles is above my current knowledge of the internals
here.

Signed-off-by: Ben Browning <bbrownin@redhat.com>

bbrowning commented on vllm/entrypoints/openai/serving_chat.py:
```diff
@@ -1393,6 +1417,8 @@ async def chat_completion_full_generator(
                 token_ids=(as_list(output.token_ids)
                            if request.return_token_ids else None),
             )
+            choice_data = self.maybe_filter_parallel_tool_calls(
+                choice_data, request)
```
I understand the spirit of this suggestion, but the reality is quite a bit more complicated than Gemini makes it sound.

For non-streaming tool call parsers, we don't even parse until after generation is finished and the parsers generally just `json.loads` the entire tool calls block at once, so there's little to no performance to gain by stopping after the first.

For streaming parsers, there would be a chance to stop parsing once we find the first. However, I don't think it's safe to assume we could stop actual generation at that point unless we can guarantee no model generates anything after the first tool call that needs to be retained.

Solving this properly to not waste any generation resources may not be generically possible across all models for streaming and non-streaming. It may be possible to solve this for a subset of models where we could add a stop token for models that have a special token output at the end of each tool call, which would let us properly stop generation after the first tool call completes.

At any rate, there's a lot to consider here and the solution in this PR is no worse performance-wise than what we do today.

bbrowning commented:
Rebased and push to resolve a conflict in serving_chat.py as it changed after I opened this PR.

chaunceyjiang commented on vllm/entrypoints/openai/serving_chat.py:
```diff
@@ -507,6 +513,31 @@ def extract_tool_call_required_streaming(
+    def maybe_filter_parallel_tool_calls(
+        self, choice: ChatCompletionResponseChoiceT, request: ChatCompletionRequest
+    ) -> ChatCompletionResponseChoiceT:
+        """Filter to first tool call only when parallel_tool_calls is False."""
```
Thanks, I think this is a reasonable change. It implements the semantics when `parallel_tool_calls` is set to False.

Could you move this function to the `vllm/entrypoints/chat_utils.py` file?

chaunceyjiang commented on docs/serving/openai_compatible_server.md:
```diff
@@ -49,7 +49,7 @@ We currently support the following OpenAI APIs:
     - *Note: `suffix` parameter is not supported.*
 - [Chat Completions API](#chat-api) (`/v1/chat/completions`)
     - Only applicable to [text generation models](../models/generative_models.md) with a [chat template](../serving/openai_compatible_server.md#chat-template).
-    - *Note: `parallel_tool_calls` and `user` parameters are ignored.*
```
I think it’s worth clarifying that this only implements the semantics for `parallel_tool_calls = False`, and not for `True`, since some models don’t natively support `parallel_tool_calls`.


@bbrowning WDYT?

bbrowning commented:
Ok, I moved the helper method out to a utils.py, added some user text explaining about `parallel_tool_calls` false vs true does, and rebased this due to a merge conflict. I re-ran the same `pytest tests/tool_use/test_parallel_tool_calls.py --models mistral llama3.2` locally and they have the same passing results after the rebase.

bbrowning commented:
It looks like this caught another Responses API test flake in function calling where gpt-oss-20b was not able to follow its harmony output format. There are likely some cleanups needed to multi-turn prompting in the Responses API based on the frequency that flakes tend to pop up there.

chaunceyjiang committed:
Merge branch 'main' into 20251003-parallel-tool-calling
